File: sdk/ts/src/adapter/Telemetry.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 RunAnywhere AI, Inc.
//
// Public adapters over PlatformBridge — mirror the Swift/Kotlin/Dart
// helper namespaces so sample-app code ports 1:1 across platforms.

import { getPlatformBridge, MissingPlatformBridgeError } from './PlatformBridge.js';

// MARK: - Telemetry

export const Telemetry = {
  /** Track a named event with optional JSON-stringified properties. */
  track(event: string, propertiesJson: string = '{}'): boolean {
    const b = getPlatformBridge();
    if (!b) throw new MissingPlatformBridgeError('Telemetry.track');
    return b.telemetryTrack(event, propertiesJson);
  },

  flush(): boolean {
    const b = getPlatformBridge();
    if (!b) throw new MissingPlatformBridgeError('Telemetry.flush');
    return b.telemetryFlush();
  },

  defaultPayloadJson(): string {
    const b = getPlatformBridge();
    if (!b) return '{}';  // pure-JS fallback for unit tests
    return b.telemetryDefaultPayloadJson();
  },
};

// MARK: - Auth

export const Auth = {
  get isAuthenticated(): boolean {
    return getPlatformBridge()?.authIsAuthenticated() ?? false;
  },

  needsRefresh(horizonSeconds: number = 60): boolean {
    return getPlatformBridge()?.authNeedsRefresh(horizonSeconds) ?? false;
  },

  get accessToken(): string  { return getPlatformBridge()?.authGetAccessToken()  ?? ''; },
  get refreshToken(): string { return getPlatformBridge()?.authGetRefreshToken() ?? ''; },
  get deviceId(): string     { return getPlatformBridge()?.authGetDeviceId()     ?? ''; },

  buildAuthenticateRequest(apiKey: string, deviceId: string): string {
    const b = getPlatformBridge();
    if (!b) throw new MissingPlatformBridgeError('Auth.buildAuthenticateRequest');
    return b.authBuildAuthenticateRequest(apiKey, deviceId);
  },

  handleAuthenticateResponse(body: string): boolean {
    const b = getPlatformBridge();
    if (!b) throw new MissingPlatformBridgeError('Auth.handleAuthenticateResponse');
    return b.authHandleAuthenticateResponse(body);
  },

  handleRefreshResponse(body: string): boolean {
    const b = getPlatformBridge();
    if (!b) throw new MissingPlatformBridgeError('Auth.handleRefreshResponse');
    return b.authHandleRefreshResponse(body);
  },

  clear(): void { getPlatformBridge()?.authClear(); },
};

// MARK: - Model helpers

export const ModelHelpers = {
  frameworkSupports(framework: string, category: string): boolean {
    return getPlatformBridge()?.modelFrameworkSupports(framework, category) ?? false;
  },

  detectFormat(urlOrPath: string): number {
    return getPlatformBridge()?.modelDetectFormat(urlOrPath) ?? 0;
  },

  inferCategory(modelId: string): number {
    return getPlatformBridge()?.modelInferCategory(modelId) ?? 0;
  },

  isArchive(urlOrPath: string): boolean {
    return getPlatformBridge()?.modelIsArchive(urlOrPath) ?? false;
  },
};

// MARK: - File integrity

export const FileIntegrity = {
  /** Hex SHA-256 of a file, or null on I/O error. */
  sha256(path: string): string | null {
    return getPlatformBridge()?.sha256File(path) ?? null;
  },

  verify(path: string, expectedHex: string): boolean {
    return getPlatformBridge()?.verifySha256(path, expectedHex) ?? false;
  },
};

// MARK: - RAG

export class RagStore {
  constructor(private handle: number) {}

  get size(): number {
    return getPlatformBridge()?.ragStoreSize(this.handle) ?? 0;
  }

  add(rowId: string, embedding: Float32Array, metadataJson: string = '{}'): boolean {
    return getPlatformBridge()?.ragStoreAdd(
      this.handle, rowId, metadataJson, embedding) ?? false;
  }

  search(query: Float32Array, topK: number = 6):
      Array<{ id: string; metadataJson: string; score: number }> {
    return getPlatformBridge()?.ragStoreSearch(this.handle, query, topK) ?? [];
  }

  close(): void {
    getPlatformBridge()?.ragStoreDestroy(this.handle);
  }

  static create(dim: number): RagStore | null {
    const b = getPlatformBridge();
    if (!b) return null;
    const h = b.ragStoreCreate(dim);
    return h === 0 ? null : new RagStore(h);
  }
}

File: sdk/kotlin/src/main/kotlin/com/runanywhere/sdk/public/Telemetry.kt
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 RunAnywhere AI, Inc.
//
// Public Kotlin telemetry/auth/model/RAG helpers backed by the
// sdk/kotlin/src/main/cpp/jni_extensions.cpp bridge. Mirror of the
// Swift adapter so sample apps get identical APIs across platforms.

package com.runanywhere.sdk.`public`

import com.runanywhere.sdk.jni.AuthNative
import com.runanywhere.sdk.jni.ModelNative
import com.runanywhere.sdk.jni.RagNative
import com.runanywhere.sdk.jni.TelemetryNative

// MARK: - Telemetry

object Telemetry {
    /** Track a named event. propertiesJson is merged into the payload. */
    @JvmStatic
    fun track(event: String, propertiesJson: String = "{}"): Boolean =
        TelemetryNative.track(event, propertiesJson) == 0

    /** Flush the pending queue to the registered HTTP uploader. */
    @JvmStatic
    fun flush(): Boolean = TelemetryNative.flush() == 0

    /** Default platform-agnostic payload (sdk version + platform). */
    @JvmStatic
    fun defaultPayloadJson(): String = TelemetryNative.defaultPayloadJson()
}

// MARK: - Auth

object Auth {
    @JvmStatic
    val isAuthenticated: Boolean
        get() = AuthNative.isAuthenticated()

    @JvmStatic
    fun needsRefresh(horizonSeconds: Int = 60): Boolean =
        AuthNative.needsRefresh(horizonSeconds)

    @JvmStatic
    val accessToken: String get() = AuthNative.getAccessToken()

    @JvmStatic
    val refreshToken: String get() = AuthNative.getRefreshToken()

    @JvmStatic
    val deviceId: String get() = AuthNative.getDeviceId()

    @JvmStatic
    fun buildAuthenticateRequest(apiKey: String, deviceId: String): String =
        AuthNative.buildAuthenticateRequest(apiKey, deviceId)

    @JvmStatic
    fun handleAuthenticateResponse(body: String): Boolean =
        AuthNative.handleAuthenticateResponse(body) == 0

    @JvmStatic
    fun clear() { AuthNative.clear() }
}

// MARK: - Model helpers

object ModelHelpers {

    @JvmStatic
    fun frameworkSupports(framework: String, category: String): Boolean =
        ModelNative.frameworkSupports(framework, category)

    @JvmStatic
    fun detectFormat(urlOrPath: String): Int = ModelNative.detectFormat(urlOrPath)

    @JvmStatic
    fun inferCategory(modelId: String): Int = ModelNative.inferCategory(modelId)

    @JvmStatic
    fun isArchive(urlOrPath: String): Boolean = ModelNative.isArchive(urlOrPath)
}

// MARK: - RAG

/** In-memory vector store backed by the C ABI `ra_rag_*` surface. */
class RagStore internal constructor(private val handle: Long) : AutoCloseable {

    val size: Int get() = RagNative.storeSize(handle)

    /** Add a row. Embedding length must match the dim passed at create. */
    fun add(rowId: String, metadataJson: String = "{}", embedding: FloatArray): Boolean =
        RagNative.storeAdd(handle, rowId, metadataJson, embedding) == 0

    /** Top-k cosine-similarity search. */
    fun search(query: FloatArray, topK: Int = 6): List<SearchHit> {
        val flat = RagNative.storeSearch(handle, query, topK)
        val hits = mutableListOf<SearchHit>()
        var i = 0
        while (i + 2 < flat.size) {
            val score = flat[i + 2].toFloatOrNull() ?: 0f
            hits += SearchHit(id = flat[i], metadataJson = flat[i + 1], score = score)
            i += 3
        }
        return hits
    }

    override fun close() {
        RagNative.storeDestroy(handle)
    }

    data class SearchHit(val id: String, val metadataJson: String, val score: Float)

    companion object {
        /** Create a new in-memory vector store. `dim` is the embedding dimension. */
        @JvmStatic
        fun create(dim: Int): RagStore? {
            val h = RagNative.storeCreate(dim)
            return if (h == 0L) null else RagStore(h)
        }
    }
}
